### PR TITLE
Do not warn when deriving `Copy` for raw pointers

### LINKED ISSUE #21272: Structs with a raw pointer warn whether or not you derive `Copy`
Damned if you do, damned if you don't:

```rust
struct foo {
bar: *mut T;
}
```

Suggests deriving `Copy`, but if `Copy` is derived, it suggests not deriving it.

Perhaps suggestion "type could implement `Copy`;" shouldn't be shown when the type has raw pointers?

### PR DESCRIPTION
**Pull request by aochagavia**
Fixes #21272 and #21296 

r? @cmr

### CODE REVIEW

**Comment by aochagavia** [comment]
@cmr Is there a way to test the lack of a warning?

**Comment by cmr** [comment]
@aochagavia you can use `#[forbid(raw_pointer_derive)]` in a `run-pass` test, and if the lint triggers the test will fail.

**Comment by aochagavia** [comment]
@cmr Thanks! I have added a test.

**Comment by aochagavia** [comment]
Any news on this?

**Comment by huonw** [comment]
Sorry, @aochagavia, lost track of this. I think it's good, just a small implementation nit.

File: src/librustc/lint/builtin.rs [hunk 1]
```diff
@@ -590,7 +590,17 @@ impl LintPass for RawPointerDerive {
             return
         }
         let did = match item.node {
-            ast::ItemImpl(..) => {
+            ast::ItemImpl(_, _, _, ref t_ref_opt, _, _) => {
+                // Deriving the Copy trait does not cause a warning
+                if let &Some(ref t_ref) = t_ref_opt {
```
Comment (huonw) [inline src/librustc/lint/builtin.rs]: Currently this code will mean that any trait called `Copy` isn't warned about, while the compiler actually has enough info to do it "perfectly", i.e. only avoid warning for the special `Copy` trait.

```rust
let trait_ref = ty::node_id_to_trait_ref(cx.tcx, t_ref.ref_id);
let def_id = ty::trait_ref_to_def_id(cx.tcx, trait_ref.def_id);
if Some(def_id) == cx.tcx.lang_items.copy_trait() {
    return
}
```

(I'm not 100% certain that will work, and I haven't compiled it.)

**Comment by huonw** [comment]
If I'm so slow again be feel to pester me.

**Comment by aochagavia** [comment]
@huonw Addressed your nit. Thanks for reviewing!

**Comment by sorccu** [comment]
Sorry to comment after closing, but should deriving Debug produce the raw_pointer_derive warning like it currently does?

**Pull request merged**